internal/packetmuxer/service.go

```diff
@@ -0,0 +1,288 @@
+		notifyTLS:            *s.NotifyTLS,
+		muxerToReliable:      *s.MuxerToReliable,
+		muxerToData:          *s.MuxerToData,
+		dataOrControlToMuxer: s.DataOrControlToMuxer,
+		muxerToNetwork:       *s.MuxerToNetwork,
+		networkToMuxer:       s.NetworkToMuxer,
+		sessionManager:       sessionManager,
+		workersManager:       workersManager,
+	}
+	workersManager.StartWorker(ws.moveUpWorker)
+	workersManager.StartWorker(ws.moveDownWorker)
+}
+
+// workersState contains the reliabletransport workers state.
+type workersState struct {
+	// logger is the logger to use
+	logger model.Logger
+
+	// hardReset is the channel posted to force a hard reset.
+	hardReset <-chan any
+
+	// notifyTLS is used to send notifications to the TLS service.
+	notifyTLS chan<- *model.Notification
+
+	// dataOrControlToMuxer is the channel for reading all the packets traveling down the stack.
+	dataOrControlToMuxer <-chan *model.Packet
+
+	// muxerToReliable is the channel for writing control packets going up the stack.
+	muxerToReliable chan<- *model.Packet
+
+	// muxerToData is the channel for writing data packets going up the stack.
+	muxerToData chan<- *model.Packet
+
+	// muxerToNetwork is the channel for writing raw packets going down the stack.
+	muxerToNetwork chan<- []byte
+
+	// networkToMuxer is the channel for reading raw packets going up the stack.
+	networkToMuxer <-chan []byte
+
+	// sessionManager manages the OpenVPN session.
+	sessionManager *session.Manager
+
+	// workersManager controls the workers lifecycle.
+	workersManager *workers.Manager
+}
+
+// moveUpWorker moves packets up the stack
+func (ws *workersState) moveUpWorker() {
+	defer func() {
+		ws.workersManager.OnWorkerDone()
+		ws.workersManager.StartShutdown()
+		ws.logger.Debug("packetmuxer: moveUpWorker: done")
+	}()
+
+	ws.logger.Debug("packetmuxer: moveUpWorker: started")
+
+	for {
+		// POSSIBLY BLOCK awaiting for incoming raw packet
+		select {
+		case rawPacket := <-ws.networkToMuxer:
+			if err := ws.handleRawPacket(rawPacket); err != nil {
+				// error already printed
+				return
+			}
+
+		case <-ws.hardReset:
+			if err := ws.startHardReset(); err != nil {
+				// error already logged
+				return
+			}
+
+		case <-ws.workersManager.ShouldShutdown():
+			return
+		}
+	}
+}
+
+// moveDownWorker moves packets down the stack
+func (ws *workersState) moveDownWorker() {
+	defer func() {
+		ws.workersManager.OnWorkerDone()
+		ws.workersManager.StartShutdown()
+		ws.logger.Debug("packetmuxer: moveDownWorker: done")
+	}()
+
+	ws.logger.Debug("packetmuxer: moveDownWorker: started")
+
+	for {
+		// POSSIBLY BLOCK on reading the packet moving down the stack
+		select {
+		case packet := <-ws.dataOrControlToMuxer:
+			// serialize the packet
+			rawPacket, err := packet.Bytes()
+			if err != nil {
+				ws.logger.Warnf("packetmuxer: cannot serialize packet: %s", err.Error())
+				continue
+			}
+
+			// While this channel send could possibly block, the [ARCHITECTURE] is
```

Comment:
i've left this comment here as a marker to consider what was our original intention; I think the whole situation might have changed a bit from where we started drafting the initial architectural stub. Whatever assumptions are valid should probably be documented in the user API, i.e., the tun interface